File: components/policy/android/java/src/org/chromium/components/policy/PolicyFeatureMap.java
// Copyright 2026 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.components.policy;

import org.jni_zero.JNINamespace;
import org.jni_zero.NativeMethods;

import org.chromium.base.FeatureMap;
import org.chromium.build.annotations.NullMarked;
import org.chromium.components.cached_flags.CachedFlag;

import java.util.List;

/** Java accessor for base/android/feature_map.h state. */
@JNINamespace("policy")
@NullMarked
public final class PolicyFeatureMap extends FeatureMap {
    private static final PolicyFeatureMap sInstance = new PolicyFeatureMap();

    // Do not instantiate this class.
    private PolicyFeatureMap() {}

    public static final CachedFlag sAndroidUseAdminsForEnterpriseInfo =
            new CachedFlag(
                    sInstance,
                    PolicyFeatures.ANDROID_USE_ADMINS_FOR_ENTERPRISE_INFO,
                    /* defaultValue= */ true,
                    /* defaultValueInTests= */ true);

    public static final List<CachedFlag> sCachedFlags = List.of(sAndroidUseAdminsForEnterpriseInfo);

    public static PolicyFeatureMap getInstance() {
        return sInstance;
    }

    public static boolean isEnabled(String featureName) {
        return getInstance().isEnabledInNative(featureName);
    }

    @Override
    protected long getNativeMap() {
        return PolicyFeatureMapJni.get().getNativeMap();
    }

    @NativeMethods
    public interface Natives {
        long getNativeMap();
    }
}

File: components/policy/core/common/android/policy_feature_map.cc
// Copyright 2026 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <string>

#include "base/android/feature_map.h"
#include "base/android/jni_android.h"
#include "base/feature_list.h"
#include "base/no_destructor.h"
#include "components/policy/core/common/features.h"

// Must come after all headers that specialize FromJniType() / ToJniType().
#include "components/policy/android/jni_headers/PolicyFeatureMap_jni.h"

namespace policy {

namespace {
// Array of features exposed through the Java PolicyFeatures API.
const base::Feature* const kFeaturesExposedToJava[] = {
    &features::kAndroidUseAdminsForEnterpriseInfo,
};

base::android::FeatureMap* GetFeatureMap() {
  static base::NoDestructor<base::android::FeatureMap> kFeatureMap(
      kFeaturesExposedToJava);
  return kFeatureMap.get();
}

}  // namespace

static int64_t JNI_PolicyFeatureMap_GetNativeMap(JNIEnv* env) {
  return reinterpret_cast<int64_t>(GetFeatureMap());
}
}  // namespace policy

DEFINE_JNI(PolicyFeatureMap)
